app/assets/javascripts/select-kit/components/categories-drop.js.es6

```diff
@@ -0,0 +1,208 @@
+import ComboBoxComponent from "select-kit/components/combo-box";
+import DiscourseURL from "discourse/lib/url";
+import { default as computed } from "ember-addons/ember-computed-decorators";
+import Category from "discourse/models/category";
+import { categoryBadgeHTML } from "discourse/helpers/category-link";
+
+export default ComboBoxComponent.extend({
+  pluginApiIdentifiers: ["categories-drop"],
+  classNames: "categories-drop",
+  classNameBindings: ["categoryStyle"],
+  categoryStyle: Ember.computed.alias("siteSettings.category_style"),
+  verticalOffset: 3,
+  headerComponent: "category-drop/category-drop-header",
+  allowAutoSelectFirst: false,
+  tagName: "li",
+  fullWidthOnMobile: true,
+  caretDownIcon: "caret-right",
+  caretUpIcon: "caret-down",
+  subCategory: false,
+  isAsync: true,
+  rowComponent: "category-row",
+
+  init() {
+    this._super(...arguments);
+
+    this.rowComponentOptions.setProperties({
+      hideParentCategory: false,
+      allowUncategorized: true,
+      category: this.subCategory || this.parentCategory,
+      countSubcategories: this.countSubcategories,
+      displayCategoryDescription: !(
+        this.currentUser &&
+        (this.currentUser.staff || this.currentUser.trust_level > 0)
+      )
+    });
+  },
+
+  @computed(
+    "shortcuts.[]",
+    "parentCategories",
+    "subCategories",
+    "parentCategory"
+  )
+  content(shortcuts, parentCategories, subCategories, parentCategory) {
+    if (parentCategory && subCategories) {
+      shortcuts = shortcuts.concat(subCategories);
+    } else {
+      shortcuts = shortcuts.concat(parentCategories);
+    }
+
+    return shortcuts;
+  },
+
+  @computed("hasSelection", "parentCategory", "subCategory", "noSubcategories")
+  shortcuts(hasSelection, parentCategory, subCategory, noSubcategories) {
+    const shortcuts = [];
+
+    if (hasSelection || (noSubcategories && subCategory)) {
+      shortcuts.push({
+        name: I18n.t("categories.all"),
+        __sk_row_type: "noopRow",
+        id: "all-categories"
+      });
+    }
+
+    if (subCategory && (hasSelection || !noSubcategories)) {
+      shortcuts.push({
+        name: this.noCategoriesLabel,
+        __sk_row_type: "noopRow",
```

Comment:
This is an unusual API. Is it something meant to be private but must be used?

Reply:
This is mostly a shortcoming of initial implementation and I had to come up with a solution. It will eventually disappear, but it gives me much flexibility for now, and also makes it very easy to track where are these special row cases so I can fix them all at once, when I have a better solution.